core/node/api_server/src/utils.rs

```diff
@@ -66,3 +77,383 @@ macro_rules! report_filter {
         ReportFilter::new($interval, &LAST_TIMESTAMP)
     }};
 }
+
+/// Fills `contract_address` for the provided transaction receipts. This field must be set
+/// only for canonical deployment transactions, i.e., txs with `to == None` for EVM contract deployment
+/// and calls to `ContractDeployer.{create, create2, createAccount, create2Account}` for EraVM contracts.
+/// Also, it must be set regardless of whether the deployment succeeded (thus e.g. we cannot rely on `ContractDeployed` events
+/// emitted by `ContractDeployer` for EraVM contracts).
+///
+/// Requires all `receipts` to be from the same L2 block.
+pub(crate) async fn fill_transaction_receipts(
+    storage: &mut Connection<'_, Core>,
+    mut receipts: Vec<ExtendedTransactionReceipt>,
+) -> Result<Vec<TransactionReceipt>, Web3Error> {
+    receipts.sort_unstable_by_key(|receipt| receipt.inner.transaction_index);
+
+    let mut filled_receipts = Vec::with_capacity(receipts.len());
+    let mut receipt_indexes_with_unknown_nonce = HashSet::new();
+    for (i, mut receipt) in receipts.into_iter().enumerate() {
+        receipt.inner.contract_address = if receipt.inner.to.is_none() {
+            // This is an EVM deployment transaction.
+            Some(deployed_address_evm_create(
+                receipt.inner.from,
+                receipt.nonce,
+            ))
```

Comment:
To highlight: Is this the correct derivation scheme? I.e., is the nonce indeed taken from the transaction (not the deployment nonce etc.)?

Reply:
Yep, for EOA transaction nonce is used (EVM doesn't have separate deployment nonce concept)